File: 310_Minimum_Height_Tree/findMinHeightTrees.py
class Solution(object):
    graph = {}
    leaves = []
    def height(self, root):
        """
        if root in self.leaves:
            return 1
        for n in self.graph[root]:
            self.height(n)
        """
        return 1 + min([self.height(n) for n in self.graph[root]]) if root not in self.leaves else 1
    
    def findMinHeightTrees(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        # Construct graph
        self.graph = {i:set() for i in xrange(n)}
        for x, y in edges:
            self.graph[x] |= {y}
            self.graph[y] |= {x}
        # got the leaves
        for node in self.graph:
            if len(self.graph[node]) == 1:
                self.leaves.append(node)

        print self.leaves
        print self.height(3)
        
        

sol = Solution()
n = 6
edges = [[0, 3], [1, 3], [2, 3], [4, 3], [5, 4]]
sol.findMinHeightTrees(n, edges)

File: 310_Minimum_Height_Tree/findMinHeightTrees.cpp
class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<pair<int, int>>& edges) {

        // corner case - no edge
        if(edges.size() == 0) { return vector<int>(1, 0); }

        vector<unordered_set<int> > graph = makeGraph(n, edges);
        vector<int> leaves, newLeaves;
        int count = n;

        for(int i = 0; i < n; i++) {
            if(graph[i].size() == 1) {
                leaves.push_back(i);
            }
        }

        while(count > 2) {
            count -= leaves.size();

            for(int i : leaves) {
                int j = *(graph[i].begin());
                graph[i].erase(j);
                graph[j].erase(i);
                if(graph[j].size() == 1) {
                    newLeaves.push_back(j);
                }
            }

            swap(leaves, newLeaves);
            newLeaves.clear();
        }

        return leaves;
    }

private:
    vector<unordered_set<int> > makeGraph(int n, vector<pair<int, int> >& edges) {
        vector<unordered_set<int> > graph(n);

        for(auto e : edges) {
            graph[e.first].insert(e.second);
            graph[e.second].insert(e.first);
        }

        return graph;
    }
};